backends/ebpf/psa/ebpfPsaDeparser.cpp

```diff
@@ -179,4 +179,125 @@ void TCIngressDeparserPSA::emitPreDeparser(CodeBuilder* builder) {
+    // clone support
+    builder->emitIndent();
+    builder->appendFormat("if (%s.clone) ", this->istd->name.name);
+    builder->blockStart();
+    builder->emitIndent();
+    builder->appendFormat(
+        "do_packet_clones(%s, &clone_session_tbl, %s.clone_session_id,"
+        " CLONE_I2E, 1);",
+        program->model.CPacketName.str(), this->istd->name.name);
+    builder->newline();
+    builder->blockEnd(true);
+}
+
+// =====================XDPIngressDeparserPSA=============================
+void XDPIngressDeparserPSA::emitPreDeparser(CodeBuilder* builder) {
+    builder->emitIndent();
+    // Perform early multicast detection; if multicast is invoked, a packet will be
+    // passed up anyway, so we can do deparsing entirely in TC.
+    // In the NTK path follow any replication path, but it has to be checked if packet is not
+    // resubmitted or multicasted because this check is not at the end of deparser.
+    cstring conditionSendToTC =
+        "if (%s->clone || %s->multicast_group != 0 ||"
+        " (!%s->drop && %s->egress_port == 0 && !%s->resubmit && %s->multicast_group == 0)) ";
+    conditionSendToTC = conditionSendToTC.replace("%s", istd->name.name);
+    builder->append(conditionSendToTC);
+    builder->blockStart();
+    builder->emitIndent();
+    builder->appendLine("struct xdp2tc_metadata xdp2tc_md = {};");
+    builder->emitIndent();
+    builder->appendFormat("xdp2tc_md.headers = *%s", this->parserHeaders->name.name);
+
+    builder->endOfStatement(true);
+    builder->emitIndent();
+    builder->appendFormat("xdp2tc_md.ostd = *%s", this->istd->name.name);
+    builder->endOfStatement(true);
+    builder->emitIndent();
+    builder->appendFormat("xdp2tc_md.packetOffsetInBits = %s", this->program->offsetVar);
+    builder->endOfStatement(true);
+    builder->append(
+        "    void *data = (void *)(long)skb->data;\n"
+        "    void *data_end = (void *)(long)skb->data_end;\n"
+        "    struct ethhdr *eth = data;\n"
+        "    if ((void *)((struct ethhdr *) eth + 1) > data_end) {\n"
+        "        return XDP_ABORTED;\n"
+        "    }\n"
+        "    xdp2tc_md.pkt_ether_type = eth->h_proto;\n"
+        "    eth->h_proto = bpf_htons(0x0800);\n");
+    if (program->options.xdp2tcMode == XDP2TC_HEAD) {
+        builder->emitIndent();
+        builder->appendFormat("int ret = bpf_xdp_adjust_head(%s, -(int)%s)",
+                              program->model.CPacketName.str(), "sizeof(struct xdp2tc_metadata)");
+        builder->endOfStatement(true);
+        builder->emitIndent();
+        builder->append("if (ret) ");
+        builder->blockStart();
+        builder->target->emitTraceMessage(builder, "Deparser: failed to push XDP2TC metadata");
+        builder->emitIndent();
+        builder->appendFormat("return %s;", builder->target->abortReturnCode().c_str());
+        builder->newline();
+        builder->blockEnd(true);
+        builder->append(
+            "    data = (void *)(long)skb->data;\n"
```

Comment:
this pattern shows up a lot, is it worthwhile to factor into a macro?

Reply:
I agree that it requires something like that.

I think rather about a function that takes a block of C code. It would split code into lines and for each line emits indent, line of code, and makes some string replace if required. Then it would be applied to a lot of places of PSA/eBPF backend. In such case I think this should be separate PR. What do you think about it?